src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -60,60 +62,68 @@ class ToplevelsWithDecorations
 class XdgDecorationManagerV1 : public wayland::XdgDecorationManagerV1
 {
 public:
-    XdgDecorationManagerV1(wl_resource* resource);
+    XdgDecorationManagerV1(wl_resource* resource, std::shared_ptr<DecorationStrategy> strategy);
 
     class Global : public wayland::XdgDecorationManagerV1::Global
     {
     public:
-        Global(wl_display* display);
+        Global(wl_display* display, std::shared_ptr<DecorationStrategy> strategy);
 
     private:
+        std::shared_ptr<DecorationStrategy> decoration_strategy;
         void bind(wl_resource* new_zxdg_decoration_manager_v1) override;
     };
 
 private:
     void get_toplevel_decoration(wl_resource* id, wl_resource* toplevel) override;
+
     ToplevelsWithDecorations toplevels_with_decorations;
+    std::shared_ptr<DecorationStrategy> decoration_strategy;
 };
 
 class XdgToplevelDecorationV1 : public wayland::XdgToplevelDecorationV1
 {
 public:
-    XdgToplevelDecorationV1(wl_resource* id, mir::frontend::XdgToplevelStable* toplevel);
+    XdgToplevelDecorationV1(
+        wl_resource* id, mir::frontend::XdgToplevelStable* toplevel, std::shared_ptr<DecorationStrategy> strategy);
 
     void set_mode(uint32_t mode) override;
     void unset_mode() override;
 
 private:
+    auto decorations_type_to_protocol_mode(DecorationStrategy::DecorationsType) -> uint32_t;
+    auto protocol_mode_to_decorations_type(uint32_t) -> DecorationStrategy::DecorationsType;
     void update_mode(uint32_t new_mode);
 
-    static uint32_t const default_mode = Mode::client_side;
-
     mir::frontend::XdgToplevelStable* toplevel;
-    uint32_t mode;
+    std::shared_ptr<DecorationStrategy> decoration_strategy;
 };
 } // namespace frontend
 } // namespace mir
 
-auto mir::frontend::create_xdg_decoration_unstable_v1(wl_display* display)
+auto mir::frontend::create_xdg_decoration_unstable_v1(wl_display* display, std::shared_ptr<DecorationStrategy> strategy)
     -> std::shared_ptr<mir::wayland::XdgDecorationManagerV1::Global>
 {
-    return std::make_shared<XdgDecorationManagerV1::Global>(display);
+    return std::make_shared<XdgDecorationManagerV1::Global>(display, strategy);
```

Comment:
Move avoids unnecessary atomic operations
```suggestion
    return std::make_shared<XdgDecorationManagerV1::Global>(display, std::move(strategy));
```